### PR TITLE
WIP 1.20.3

### PR DESCRIPTION
**Pull request by electronicboy**

### CODE REVIEW

File: proxy/src/main/java/com/velocitypowered/proxy/connection/client/ConnectedPlayer.java [hunk 1]
```diff
@@ -1015,22 +1017,36 @@ private void tickResourcePackQueue() {
...
   @Override
+  @Deprecated
   public @Nullable ResourcePackInfo getPendingResourcePack() {
+    //TODO which resource pack should be returned here?
     return pendingResourcePack;
   }
 
+  @Override
+  public Collection<ResourcePackInfo> getAppliedResourcePacks() {
+    //TODO
```
Comment (Timongcraft) [inline proxy/src/main/java/com/velocitypowered/proxy/connection/client/ConnectedPlayer.java:1042]: Shouln't this be finished before a merge?

File: proxy/src/main/java/com/velocitypowered/proxy/connection/client/ConnectedPlayer.java [hunk 2]
```diff
@@ -1015,22 +1017,36 @@ private void tickResourcePackQueue() {
...
+  }
+
+  @Override
+  public Collection<ResourcePackInfo> getPendingResourcePacks() {
+    //TODO
```
Comment (Timongcraft) [inline proxy/src/main/java/com/velocitypowered/proxy/connection/client/ConnectedPlayer.java:1047]: Shouln't this be finished before a merge?

File: proxy/src/main/java/com/velocitypowered/proxy/connection/backend/BackendPlaySessionHandler.java [hunk 3]
```diff
@@ -213,6 +216,11 @@ public boolean handle(ResourcePackRequest packet) {
     return true;
   }
 
+  @Override
+  public boolean handle(RemoveResourcePack packet) {
+    //TODO
```
Comment (Timongcraft) [inline proxy/src/main/java/com/velocitypowered/proxy/connection/backend/BackendPlaySessionHandler.java:221]: Shouln't this be done before a merge?

Comment (4drian3d) [reply proxy/src/main/java/com/velocitypowered/proxy/connection/backend/BackendPlaySessionHandler.java:221]: Obviously, this pull request is still under development, it is not necessary to mention the remaining TODOs

File: proxy/src/main/java/com/velocitypowered/proxy/connection/client/ConnectedPlayer.java (hunk 1 above)
Comment (electronicboy) [reply proxy/src/main/java/com/velocitypowered/proxy/connection/client/ConnectedPlayer.java:1042]: Yes, that's the entire point it was converted into a PR.

**Comment by MrPowerGamerBR** [comment]
One thing that I have noticed while testing (I have replaced all the `//TODO` with dummy return values) is that I cannot login to the server
```
[12:29:44 ERROR]: [server connection] MrPowerGamerBR -> sparklypower_survival: exception encountered in com.velocitypowered.proxy.connection.backend.BackendPlaySessionHandler@62052405
java.lang.NullPointerException: Cannot invoke "com.velocitypowered.proxy.protocol.packet.chat.ComponentHolder.getComponent()" because the return value of "com.velocitypowered.proxy.protocol.packet.UpsertPlayerInfo$Entry.getDisplayName()" is null
        at com.velocitypowered.proxy.tablist.VelocityTabList.processUpsert(VelocityTabList.java:255) ~[velocity-proxy-3.2.0-sparklyvelocity-SNAPSHOT-all.jar:3.2.0-sparklyvelocity-SNAPSHOT (git-ba58425d)]
        at com.velocitypowered.proxy.tablist.VelocityTabList.processUpdate(VelocityTabList.java:205) ~[velocity-proxy-3.2.0-sparklyvelocity-SNAPSHOT-all.jar:3.2.0-sparklyvelocity-SNAPSHOT (git-ba58425d)]
        at com.velocitypowered.proxy.connection.backend.BackendPlaySessionHandler.handle(BackendPlaySessionHandler.java:281) ~[velocity-proxy-3.2.0-sparklyvelocity-SNAPSHOT-all.jar:3.2.0-sparklyvelocity-SNAPSHOT (git-ba58425d)]
        at com.velocitypowered.proxy.protocol.packet.UpsertPlayerInfo.handle(UpsertPlayerInfo.java:133) ~[velocity-proxy-3.2.0-sparklyvelocity-SNAPSHOT-all.jar:3.2.0-sparklyvelocity-SNAPSHOT (git-ba58425d)]
        at com.velocitypowered.proxy.connection.MinecraftConnection.channelRead(MinecraftConnection.java:153) ~[velocity-proxy-3.2.0-sparklyvelocity-SNAPSHOT-all.jar:3.2.0-sparklyvelocity-SNAPSHOT (git-ba58425d)]
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:444) ~[velocity-proxy-3.2.0-sparklyvelocity-SNAPSHOT-all.jar:3.2.0-sparklyvelocity-SNAPSHOT (git-ba58425d)]
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420) ~[velocity-proxy-3.2.0-sparklyvelocity-SNAPSHOT-all.jar:3.2.0-sparklyvelocity-SNAPSHOT (git-ba58425d)]
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412) ~[velocity-proxy-3.2.0-sparklyvelocity-SNAPSHOT-all.jar:3.2.0-sparklyvelocity-SNAPSHOT (git-ba58425d)]
        at com.velocitypowered.proxy.protocol.netty.AutoReadHolderHandler.channelRead(AutoReadHolderHandler.java:57) ~[velocity-proxy-3.2.0-sparklyvelocity-SNAPSHOT-all.jar:3.2.0-sparklyvelocity-SNAPSHOT (git-ba58425d)]
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:442) ~[velocity-proxy-3.2.0-sparklyvelocity-SNAPSHOT-all.jar:3.2.0-sparklyvelocity-SNAPSHOT (git-ba58425d)]
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420) ~[velocity-proxy-3.2.0-sparklyvelocity-SNAPSHOT-all.jar:3.2.0-sparklyvelocity-SNAPSHOT (git-ba58425d)]
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412) ~[velocity-proxy-3.2.0-sparklyvelocity-SNAPSHOT-all.jar:3.2.0-sparklyvelocity-SNAPSHOT (git-ba58425d)]
        at com.velocitypowered.proxy.protocol.netty.MinecraftDecoder.tryDecode(MinecraftDecoder.java:92) ~[velocity-proxy-3.2.0-sparklyvelocity-SNAPSHOT-all.jar:3.2.0-sparklyvelocity-SNAPSHOT (git-ba58425d)]
        at com.velocitypowered.proxy.protocol.netty.MinecraftDecoder.channelRead(MinecraftDecoder.java:61) ~[velocity-proxy-3.2.0-sparklyvelocity-SNAPSHOT-all.jar:3.2.0-sparklyvelocity-SNAPSHOT (git-ba58425d)]
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:444) ~[velocity-proxy-3.2.0-sparklyvelocity-SNAPSHOT-all.jar:3.2.0-sparklyvelocity-SNAPSHOT (git-ba58425d)]
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420) ~[velocity-proxy-3.2.0-sparklyvelocity-SNAPSHOT-all.jar:3.2.0-sparklyvelocity-SNAPSHOT (git-ba58425d)]
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412) ~[velocity-proxy-3.2.0-sparklyvelocity-SNAPSHOT-all.jar:3.2.0-sparklyvelocity-SNAPSHOT (git-ba58425d)]
        at io.netty.handler.timeout.IdleStateHandler.channelRead(IdleStateHandler.java:286) ~[velocity-proxy-3.2.0-sparklyvelocity-SNAPSHOT-all.jar:3.2.0-sparklyvelocity-SNAPSHOT (git-ba58425d)]
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:442) ~[velocity-proxy-3.2.0-sparklyvelocity-SNAPSHOT-all.jar:3.2.0-sparklyvelocity-SNAPSHOT (git-ba58425d)]
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420) ~[velocity-proxy-3.2.0-sparklyvelocity-SNAPSHOT-all.jar:3.2.0-sparklyvelocity-SNAPSHOT (git-ba58425d)]
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412) ~[velocity-proxy-3.2.0-sparklyvelocity-SNAPSHOT-all.jar:3.2.0-sparklyvelocity-SNAPSHOT (git-ba58425d)]
        at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:346) ~[velocity-proxy-3.2.0-sparklyvelocity-SNAPSHOT-all.jar:3.2.0-sparklyvelocity-SNAPSHOT (git-ba58425d)]
        at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:318) ~[velocity-proxy-3.2.0-sparklyvelocity-SNAPSHOT-all.jar:3.2.0-sparklyvelocity-SNAPSHOT (git-ba58425d)]
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:444) ~[velocity-proxy-3.2.0-sparklyvelocity-SNAPSHOT-all.jar:3.2.0-sparklyvelocity-SNAPSHOT (git-ba58425d)]
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420) ~[velocity-proxy-3.2.0-sparklyvelocity-SNAPSHOT-all.jar:3.2.0-sparklyvelocity-SNAPSHOT (git-ba58425d)]
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412) ~[velocity-proxy-3.2.0-sparklyvelocity-SNAPSHOT-all.jar:3.2.0-sparklyvelocity-SNAPSHOT (git-ba58425d)]
        at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410) ~[velocity-proxy-3.2.0-sparklyvelocity-SNAPSHOT-all.jar:3.2.0-sparklyvelocity-SNAPSHOT (git-ba58425d)]
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:440) ~[velocity-proxy-3.2.0-sparklyvelocity-SNAPSHOT-all.jar:3.2.0-sparklyvelocity-SNAPSHOT (git-ba58425d)]
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420) ~[velocity-proxy-3.2.0-sparklyvelocity-SNAPSHOT-all.jar:3.2.0-sparklyvelocity-SNAPSHOT (git-ba58425d)]
        at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919) ~[velocity-proxy-3.2.0-sparklyvelocity-SNAPSHOT-all.jar:3.2.0-sparklyvelocity-SNAPSHOT (git-ba58425d)]
        at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:166) ~[velocity-proxy-3.2.0-sparklyvelocity-SNAPSHOT-all.jar:3.2.0-sparklyvelocity-SNAPSHOT (git-ba58425d)]
        at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:788) ~[velocity-proxy-3.2.0-sparklyvelocity-SNAPSHOT-all.jar:3.2.0-sparklyvelocity-SNAPSHOT (git-ba58425d)]
        at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:724) ~[velocity-proxy-3.2.0-sparklyvelocity-SNAPSHOT-all.jar:3.2.0-sparklyvelocity-SNAPSHOT (git-ba58425d)]
        at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:650) ~[velocity-proxy-3.2.0-sparklyvelocity-SNAPSHOT-all.jar:3.2.0-sparklyvelocity-SNAPSHOT (git-ba58425d)]
        at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:562) ~[velocity-proxy-3.2.0-sparklyvelocity-SNAPSHOT-all.jar:3.2.0-sparklyvelocity-SNAPSHOT (git-ba58425d)]
        at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:997) ~[velocity-proxy-3.2.0-sparklyvelocity-SNAPSHOT-all.jar:3.2.0-sparklyvelocity-SNAPSHOT (git-ba58425d)]
        at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74) ~[velocity-proxy-3.2.0-sparklyvelocity-SNAPSHOT-all.jar:3.2.0-sparklyvelocity-SNAPSHOT (git-ba58425d)]
        at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30) ~[velocity-proxy-3.2.0-sparklyvelocity-SNAPSHOT-all.jar:3.2.0-sparklyvelocity-SNAPSHOT (git-ba58425d)]
        at java.lang.Thread.run(Thread.java:833) [?:?]
```

What I have noticed is that `setDisplayNameWithoutUpdate` *is* nullable, but `getDisplayName()` can be nullable too `currentEntry.setDisplayNameWithoutUpdate(entry.getDisplayName().getComponent());`

**Comment by electronicboy** [comment]
This was an early WIP branch that I converted to a PR, but looked like somebody else was working on it as I was in my death bed, and somebody else has opened a PR, so, eer

**Comment by MrPowerGamerBR** [comment]
> This was an early WIP branch that I converted to a PR, but looked like somebody else was working on it as I was in my death bed, and somebody else has opened a PR, so, eer

I was testing this PR out before I noticed pkt's #1150 PR 😭

**Pull request closed without merging**